rust/signer/src/ffi_types.rs

```diff
@@ -0,0 +1,110 @@
+use crate::UniffiCustomTypeConverter;
+pub use definitions::{
+    crypto::Encryption,
+    history::{
+        Event, IdentityHistory, MetaValuesDisplay, MetaValuesExport, NetworkSpecsDisplay,
+        NetworkSpecsExport, NetworkVerifierDisplay, SignDisplay, SignMessageDisplay, TypesDisplay,
+        TypesExport,
+    },
+    metadata::MetaValues,
+    navigation::*,
+    network_specs::{
+        NetworkSpecs, NetworkSpecsToSend, ValidCurrentVerifier, Verifier, VerifierValue,
+    },
+};
+pub use navigator::Action;
+
+use std::convert::{TryFrom, TryInto};
+
+pub type NetworkSpecsKey = definitions::keyring::NetworkSpecsKey;
+
+impl UniffiCustomTypeConverter for NetworkSpecsKey {
+    type Builtin = String;
+
+    fn into_custom(hex_line: Self::Builtin) -> uniffi::Result<Self> {
+        Self::from_hex(&hex_line).map_err(|_| anyhow::Error::msg("network specks key"))
+    }
+
+    fn from_custom(obj: Self) -> Self::Builtin {
+        hex::encode(obj.key())
+    }
+}
+
+pub type Ed25519Public = sp_core::ed25519::Public;
+pub type Sr25519Public = sp_core::sr25519::Public;
+pub type EcdsaPublic = sp_core::ecdsa::Public;
+pub type MultiSigner = sp_runtime::MultiSigner;
+pub type H256 = sp_core::H256;
+
+impl UniffiCustomTypeConverter for sp_runtime::MultiSigner {
+    type Builtin = String;
+
+    fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+        Ok(serde_json::from_str(&val)?)
```

Comment:
This should be done like in `db_handling::helpers::get_general_verifier` with separate fields for pubkey, identicon and encryption. With default upstream serde implementation, this results in base58 representation for nonsence network; thus history starts with weird things.

Reply:
I've added a hack that sends `vec[key, encryption]` as strings for now. This should be properly encoded as `MVerifierDetails` I guess but for now I have failed to make it work with uniffi machinery